src/scsmodule.c

```diff
@@ -431,12 +447,35 @@ static PyObject *csolve(PyObject *self, PyObject *args, PyObject *kwargs) {
     d->stgs->warm_start |= get_warm_start("y", &(sol.y), d->m, warm);
     d->stgs->warm_start |= get_warm_start("s", &(sol.s), d->m, warm);
   }
+
+#ifdef PYTHON_LINSYS
+  if (!PyCallable_Check(solve_lin_sys_cb)) {
+    PyErr_SetString(PyExc_ValueError, "solve_lin_sys_cb not a valid callback");
+    return SCS_NULL;
+  }
+
+  if (!PyCallable_Check(accum_by_a_cb)) {
+    PyErr_SetString(PyExc_ValueError, "accum_by_a_cb not a valid callback");
+    return SCS_NULL;
+  }
+
+  if (!PyCallable_Check(accum_by_atrans_cb)) {
+    PyErr_SetString(PyExc_ValueError, "accum_by_atrans_cb not a valid callback");
+    return SCS_NULL;
+  }
+#endif
+
+
+#ifndef PYTHON_LINSYS
   /* release the GIL */
   Py_BEGIN_ALLOW_THREADS;
```

Comment:
We could release the GIL here and then reacquire once we enter each callback (then release at the end of each callback), that would probably be better for multi-threading.

Reply:
Hmm, I thought about this earlier. This would require sharing the python thread state from this part with the linsys calls. I think the easiest way of adding this would be to have a global `_save` variable in `scs-module.c` that `private.c` modifies as it's acquiring/releasing the GIL around the Python callbacks. What do you think?

Ref: https://docs.python.org/3/c-api/init.html#c.Py_BEGIN_ALLOW_THREADS

Reply:
Hmm, I just tried doing this and am getting a segfault and am not sure what else to try. I put this around the main `scs` call:

```python
/* release the GIL */
scs_thread_state = PyEval_SaveThread();
/* Solve! */
scs(d, k, &sol, &info);
/* reacquire the GIL */
PyEval_RestoreThread(scs_thread_state);
```

And this around all of the callbacks:

```python
PyEval_RestoreThread(scs_thread_state);
PyObject_CallObject(scs_solve_lin_sys_cb, arglist);
scs_thread_state = PyEval_SaveThread();
```

Reply:
Weird, ok let's just go with what you have for now then.